File: Java/src/test/java/com/gildedrose/GildedRoseTest.java
package com.gildedrose;

import com.gildedrose.item.Item;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class GildedRoseTest {

    @Test
    void item() {
        Item[] items = new Item[]{
            new Item("foo", 0, 1),
            new Item("bar", 0, 2)
        };
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals("foo", app.items[0].name);
        assertEquals(-1, app.items[0].sellIn);
        assertEquals(0, app.items[0].quality);
        assertEquals("foo, -1, 0", app.items[0].toString());

        assertEquals("bar", app.items[1].name);
        assertEquals(-1, app.items[1].sellIn);
        assertEquals(0, app.items[1].quality);
        assertEquals("bar, -1, 0", items[1].toString());
    }

    @Test
    void itemLegendary() {
        Item[] items = new Item[]{new Item("Sulfuras, Hand of Ragnaros", 0, 1)};

        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals("Sulfuras, Hand of Ragnaros", app.items[0].name);
        assertEquals(1, app.items[0].quality);
        assertEquals(1, app.items[0].quality);
    }

    @Test
    void itemAgedBrie() {
        Item[] items = new Item[]{new Item("Aged Brie", 0, 0)};

        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals("Aged Brie", app.items[0].name);
        assertEquals(2, app.items[0].quality);
        assertEquals(-1, app.items[0].sellIn);
    }

    @Test
    void itemBackstagePass() {
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 0, 0)};

        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        assertEquals(0, app.items[0].quality);
        assertEquals(-1, app.items[0].sellIn);
    }
}

File: go/gildedrose/gildedrose_test.go
package gildedrose_test

import (
	"testing"

	"github.com/emilybache/gildedrose-refactoring-kata/gildedrose"
)

func Test_Foo(t *testing.T) {
	var items = []*gildedrose.Item{
		{"foo", 0, 0},
		{"bar", 1, 2},
		{"baz", -1, 2},
		{"qux", 1, -2},
		{"qux", 1, 55},
	}

	expected := []*gildedrose.Item{
		{"foo", -1, 0},
		{"bar", 0, 1},
		{"baz", -2, 0},
		{"qux", 0, -2},
		{"qux", 0, 54},
	}

	if len(items) != len(expected) {
		t.Errorf("expected items should have %d items", len(expected))
		return
	}

	gildedrose.UpdateQuality(items)

	for i, item := range items {
		if item.Name != expected[i].Name {
			t.Errorf("Name: Expected %s but got %s ", expected[i].Name, item.Name)
		}
		if item.SellIn != expected[i].SellIn {
			t.Errorf("SellIn: Expected %d but got %d ", expected[i].SellIn, item.SellIn)
		}
		if item.Quality != expected[i].Quality {
			t.Errorf("Quality: Expected %d but got %d ", expected[i].Quality, item.Quality)
		}
	}
}

func Test_AgedBrie(t *testing.T) {
	var items = []*gildedrose.Item{
		{"Aged Brie", -1, 2},
		{"Aged Brie", 0, 0},
		{"Aged Brie", 1, -1},
		{"Aged Brie", 10, 30},
		{"Aged Brie", 10, 55},
	}

	expected := []*gildedrose.Item{
		{"Aged Brie", -2, 4},
		{"Aged Brie", -1, 2},
		{"Aged Brie", 0, 0},
		{"Aged Brie", 9, 31},
		{"Aged Brie", 9, 55},
	}

	if len(items) != len(expected) {
		t.Errorf("expected items should have %d items", len(expected))
		return
	}

	gildedrose.UpdateQuality(items)

	for i, item := range items {
		if item.Name != expected[i].Name {
			t.Errorf("Name: Expected %s but got %s ", expected[i].Name, item.Name)
		}
		if item.SellIn != expected[i].SellIn {
			t.Errorf("SellIn: Expected %d but got %d ", expected[i].SellIn, item.SellIn)
		}
		if item.Quality != expected[i].Quality {
			t.Errorf("Quality: Expected %d but got %d ", expected[i].Quality, item.Quality)
		}
	}
}

func Test_BackstagePass(t *testing.T) {
	var items = []*gildedrose.Item{
		{"Backstage passes to a TAFKAL80ETC concert", 0, 0},
		{"Backstage passes to a TAFKAL80ETC concert", 1, 0},
		{"Backstage passes to a TAFKAL80ETC concert", 0, 1},
		{"Backstage passes to a TAFKAL80ETC concert", 1, -1},
		{"Backstage passes to a TAFKAL80ETC concert", -1, 1},
		{"Backstage passes to a TAFKAL80ETC concert", -1, 50},
	}

	expected := []*gildedrose.Item{
		{"Backstage passes to a TAFKAL80ETC concert", -1, 0},
		{"Backstage passes to a TAFKAL80ETC concert", 0, 3},
		{"Backstage passes to a TAFKAL80ETC concert", -1, 0},
		{"Backstage passes to a TAFKAL80ETC concert", 0, 2},
		{"Backstage passes to a TAFKAL80ETC concert", -2, 0},
		{"Backstage passes to a TAFKAL80ETC concert", -2, 0},
	}

	if len(items) != len(expected) {
		t.Errorf("expected items should have %d items", len(expected))
		return
	}

	gildedrose.UpdateQuality(items)

	for i, item := range items {
		if item.Name != expected[i].Name {
			t.Errorf("Name: Expected %s but got %s ", expected[i].Name, item.Name)
		}
		if item.SellIn != expected[i].SellIn {
			t.Errorf("SellIn: Expected %d but got %d ", expected[i].SellIn, item.SellIn)
		}
		if item.Quality != expected[i].Quality {
			t.Errorf("Quality: Expected %d but got %d ", expected[i].Quality, item.Quality)
		}
	}
}

func Test_Sulfuras(t *testing.T) {
	var items = []*gildedrose.Item{
		{"Sulfuras, Hand of Ragnaros", 0, 0},
		{"Sulfuras, Hand of Ragnaros", 1, 0},
		{"Sulfuras, Hand of Ragnaros", 0, 1},
		{"Sulfuras, Hand of Ragnaros", 1, -1},
		{"Sulfuras, Hand of Ragnaros", -1, 1},
		{"Sulfuras, Hand of Ragnaros", -1, 50},
	}

	expected := []*gildedrose.Item{
		{"Sulfuras, Hand of Ragnaros", 0, 0},
		{"Sulfuras, Hand of Ragnaros", 1, 0},
		{"Sulfuras, Hand of Ragnaros", 0, 1},
		{"Sulfuras, Hand of Ragnaros", 1, -1},
		{"Sulfuras, Hand of Ragnaros", -1, 1},
		{"Sulfuras, Hand of Ragnaros", -1, 50},
	}

	if len(items) != len(expected) {
		t.Errorf("expected items should have %d items", len(expected))
		return
	}

	gildedrose.UpdateQuality(items)

	for i, item := range items {
		if item.Name != expected[i].Name {
			t.Errorf("Name: Expected %s but got %s ", expected[i].Name, item.Name)
		}
		if item.SellIn != expected[i].SellIn {
			t.Errorf("SellIn: Expected %d but got %d ", expected[i].SellIn, item.SellIn)
		}
		if item.Quality != expected[i].Quality {
			t.Errorf("Quality: Expected %d but got %d ", expected[i].Quality, item.Quality)
		}
	}
}
